accepted/2025/rid-specific-tool-packages.md

```diff
@@ -0,0 +1,91 @@
+
+## Entry points
+
+Recall that the tool manifest file has a section that looks like this:
+
+```xml
+<Command Name="dotnet-say" EntryPoint="dotnet-say.dll" Runner="dotnet" />
+```
+
+Currently, the only supported value for `Runner` is `dotnet`, which means that the `EntryPoint` is a DLL that can be launched with the `dotnet` executable.  When a global tool is installed, the .NET SDK will create a version of the .NET apphost (called a shim) in the global tools folder.  This shim has a filename matching the tool command name so that the tool can be launched from the path.  The shim also has the relative path to the tool entry point embedded in it so that it can launch the correct tool.  Tools can also include their own shims so that they can be properly signed.
+
+To support NativeAOT .NET apps (or non .NET apps), we should support an additional `executable` runner for executables that can be launched natively by the operating system.  On Unix-like operating systems, we can create a symlink in the global tools folder to the tool entry point executable.  On Windows, we may create a batch file that launches the entry point executable.  If this has drawbacks we might create a new type of shim/launcher.
+
+An alternative would be to copy the whole entry point executable to the global tools folder.  However, there might be other files in the same folder that the tool depends on, so we don't currently plan to do this.
+
+## Tool manifest format version
+
+The root element of the tool manifest has a Version attribute.  Currently this is set to 1.  When creating a tool manifest with a `RuntimeIdentifierPackages` element or a command Runner type of `executable`, we should set the Version to 2, so that prior .NET SDKs that don't support those features will error out gracefully when trying to install the tool.  If the tool manifest doesn't use these new features, the version should still be set to 1 so that the tool can be consumed by previous SDKs.
+
+## Production
+
+A tool project will be able to specify the runtime identifiers to create RID-specific packages for using the `ToolPackageRuntimeIdentifier` item.  For example:
+
+```xml
+<ItemGroup>
+	<ToolPackageRuntimeIdentifier Include="win-x64"/>
+	<ToolPackageRuntimeIdentifier Include="linux-x64"/>
+</ItemGroup>
+```
+
+If `ToolPackageRuntimeIdentifier` is non-empty, then packing the project without a RuntimeIdentifier will create the primary package, and packing the project with a RuntimeIdentifier will create the corresponding RID-specific package.  If feasible, we could do an automatic inner pack of the ToolPackageRuntimeIdentifiers before creating the primary tool package.
+
+The package name for the RID-specific packages will be `<ToolPackageName>.<RuntimeIdentifier>`.  If not specified, the version number of the RID-specific package will be the same as that of the primary package.  `Version` metadata on the `ToolPackageRuntimeIdentifier` item can be used to set the RID-specific package version.
```

Comment:
Is `Version` a version or version range? Seems like version. What's the error experience when the RID-specific package cannot be found?

Reply:
It's an exact version.  If that version isn't found then it should error out, probably in a similar way to what would happen if the tool couldn't be found.